legacy/src/app/directives/boot_images.js

```diff
@@ -322,23 +366,79 @@ export function maasBootImages(
       return filtered;
     };
 
-    // Return true if the source has this selected.
-    $scope.isSelected = function (type, obj) {
-      return $scope.source.selections[type].indexOf(obj) >= 0;
+    // Return true if the release/architecture combination is selected.
+    $scope.isOSSelected = (release, archName) => {
+      const osystem = $scope.source.selections.osystems.find(
+        (osystem) => osystem.release === release
+      );
+      return osystem && osystem.arches && osystem.arches.includes(archName);
     };
 
-    // Toggle the selection of the release or the architecture.
-    $scope.toggleSelection = function (type, obj) {
-      var idx = $scope.source.selections[type].indexOf(obj);
-      if (idx === -1) {
-        $scope.source.selections[type].push(obj);
+    // Toggle the selection of the release/architecture combination.
+    $scope.toggleSelectedOS = (releaseName, archName) => {
+      const selectedOSs = $scope.source.selections.osystems;
+      let newSelected = [];
+      const osystemInSelected = selectedOSs.find(
+        (osystem) => osystem.release === releaseName
+      );
+      if (osystemInSelected) {
+        if (osystemInSelected.arches.includes(archName)) {
+          if (osystemInSelected.arches.length === 1) {
```

Comment:
Perhaps consider factoring this out if you can, three levels of `if else` statements are a bit hard to read. If you moved this logic to a function outside of toggleSelectedOS, perhaps you could return `newSelected` early which might help with readability, e.g. something like:

```
if (!osystemInSelected) {
        return newSelected = [
          ...selectedOSs,
          { osystem: "ubuntu", release: releaseName, arches: [archName] },
        ];
}
.... etc
```

Reply:
Good call - I think I was lost in the sauce a bit when writing it. I've factored out the newSelected part to its own function so it returns out earlier, and flattened the ifs.